PythonVisualizations/PriorityQueue.py

```diff
@@ -0,0 +1,367 @@
+import random
+from tkinter import *
+
+try:
+    from drawable import *
+    from VisualizationApp import *
+except ModuleNotFoundError:
+    from .drawable import *
+    from .VisualizationApp import *
+
+
+class PriorityQueue(VisualizationApp):
+    CELL_SIZE = 50
+    CELL_BORDER = 2
+    CELL_BORDER_COLOR = 'black'
+    ARRAY_X0 = 100
+    ARRAY_Y0 = 100
+
+    nextColor = 0
+
+    def __init__(self, size=12, title="Priority Queue", **kwargs):
+        super().__init__(title=title, **kwargs)
+        self.size = size
+        self.title = title
+        self.list = [None]*self.size
+        self.nItems = 0
+        self.buttons = self.makeButtons()
+        self.index = self.createIndex(self.nItems-1, 'front', level=-1) # indicate priority
+        
+        self.display()
+
+    def __str__(self):
+        return str(self.list)
+
+    # ARRAY FUNCTIONALITY
+
+    def createIndex(self, index,  # cell
+                    name=None,  # with an optional name label
+                    level=1,  # at a particular level away from the cells
+                    color=None):  # (negative are below)
+        if not color: color = self.VARIABLE_COLOR
+
+        cell_coords = self.cellCoords(index)
+        cell_center = self.cellCenter(index)
+        level_spacing = self.VARIABLE_FONT[1]
+        x = cell_center[0]
+        if level > 0:
+            y0 = cell_coords[1] - self.CELL_SIZE * 3 // 5 - level * level_spacing
+            y1 = cell_coords[1] - self.CELL_SIZE * 3 // 10
+        else:
+            y0 = cell_coords[3] + self.CELL_SIZE * 3 // 5 - level * level_spacing
+            y1 = cell_coords[3] + self.CELL_SIZE * 3 // 10
+        arrow = self.canvas.create_line(
+            x, y0, x, y1, arrow="last", fill=color)
+        if name:
+            label = self.canvas.create_text(
+                x + 2, y0, text=name, anchor=SW if level > 0 else NW,
+                font=self.VARIABLE_FONT, fill=color)
+        return (arrow, label) if name else (arrow,)
+
+    def insert(self, val):
+        callEnviron = self.createCallEnvironment()
+        self.startAnimations()
+        j = self.nItems - 1  # Start at front
+
+        indexJ = self.createIndex(j, 'j', level=-2)
+        callEnviron |= set(indexJ)
+
+        #self.list[j] = drawable(None)
+        self.list[self.nItems] = drawable(None)
+        
+        while j >= 0 and val >= self.list[j].val:  # Move bigger items right
+
+            self.moveItemsBy(indexJ, (-self.CELL_SIZE, 0), sleepTime=0.1)  # Move "j" arrow
+            self.wait(1)
+            
+            self.assignElement(j+1, j, callEnviron)
+            self.list[j+1].val = self.list[j].val
+            j -= 1
+            self.wait(1)
+
+        self.nItems += 1
+            
```

Comment:
There are several problems with this loop to shift cells over.  Here's what I think the loop should look like:
```
        self.wait(1)
            
        while j >= 0 and val >= self.list[j].val:  # Move bigger items right

            self.assignElement(j, j+1, callEnviron)
            self.list[j+1].val = self.list[j].val
            j -= 1
            self.moveItemsBy(indexJ, (-self.CELL_SIZE, 0), sleepTime=0.1)  # Move "j" arrow
            self.wait(1)
```

Here's why:
 * Moving the first call to `wait()` outside the loop makes the animation pause; that is to show it's taking time to test the loop condition
 * Moving `indexJ` happens after shifting the item over in the textbook code, so we need to do it in the same order here.
 * The arguments for `fromIndex` and `toIndex` were swapped.  We are moving the item at `j`, to cell `j+1` to match what the textbook code does.
 * When we decrement `j`, that's when we move `indexJ`
 * The call to `wait()` at the end of the loop will make the animation pause while it evaluates the loop condition

Furthermore incrementing the `nItems` value at line 82 doesn't happen until after the assignment of the new item to cell j, so delete line 82.  We'll put it back in below.